File: core/metrics.py
"""
Módulo de Cálculo de Métricas

Calcula todas as métricas de performance de backtesting
"""

import numpy as np
import pandas as pd
from typing import List, Dict, Optional


class MetricsCalculator:
    """
    Calcula métricas de performance de trading
    """
    
    def __init__(self, risk_free_rate: float = 0.0):
        """
        Args:
            risk_free_rate: Taxa livre de risco anual (ex: 0.1 = 10%)
        """
        self.risk_free_rate = risk_free_rate
    
    def calculate_all(self, trades: List[Dict]) -> Dict:
        """
        Calcula todas as métricas
        
        Args:
            trades: Lista de trades com 'pnl', 'type', 'exit_reason', etc.
        
        Returns:
            Dict com todas as métricas
        """
        if not trades:
            return self._empty_metrics()
        
        trades_df = pd.DataFrame(trades)
        
        # Retornos
        returns = trades_df['pnl'].values
        
        # Métricas básicas
        total_return = returns.sum()
        total_trades = len(trades)
        
        winning_trades = trades_df[trades_df['pnl'] > 0]
        losing_trades = trades_df[trades_df['pnl'] < 0]
        
        win_rate = len(winning_trades) / total_trades if total_trades > 0 else 0
        
        avg_win = winning_trades['pnl'].mean() if len(winning_trades) > 0 else 0
        avg_loss = losing_trades['pnl'].mean() if len(losing_trades) > 0 else 0
        
        # Profit Factor
        gross_profit = winning_trades['pnl'].sum() if len(winning_trades) > 0 else 0
        gross_loss = abs(losing_trades['pnl'].sum()) if len(losing_trades) > 0 else 0
        profit_factor = gross_profit / gross_loss if gross_loss != 0 else 0
        
        # Sharpe Ratio
        sharpe = self._calculate_sharpe(returns)
        
        # Sortino Ratio
        sortino = self._calculate_sortino(returns)
        
        # Max Drawdown
        equity_curve = np.cumsum(returns)
        max_dd, max_dd_pct = self._calculate_max_drawdown(equity_curve)
        
        # Consecutive wins/losses
        max_consecutive_wins = self._max_consecutive(trades_df, 'win')
        max_consecutive_losses = self._max_consecutive(trades_df, 'loss')
        
        # Expectancy
        expectancy = (win_rate * avg_win) + ((1 - win_rate) * avg_loss)
        
        # LONG vs SHORT
        long_trades = trades_df[trades_df['type'] == 'LONG']
        short_trades = trades_df[trades_df['type'] == 'SHORT']
        
        # SL vs TP
        sl_trades = trades_df[trades_df.get('exit_reason', '') == 'SL']
        tp_trades = trades_df[trades_df.get('exit_reason', '') == 'TP']
        
        return {
            # Retorno
            'total_return': total_return,
            'total_return_pct': (total_return / 10000) * 100 if total_return != 0 else 0,
            
            # Trades
            'total_trades': total_trades,
            'winning_trades': len(winning_trades),
            'losing_trades': len(losing_trades),
            'win_rate': win_rate,
            
            # Médias
            'avg_win': avg_win,
            'avg_loss': avg_loss,
            'avg_trade': returns.mean(),
            
            # Performance
            'sharpe_ratio': sharpe,
            'sortino_ratio': sortino,
            'profit_factor': profit_factor,
            'expectancy': expectancy,
            
            # Drawdown
            'max_drawdown': max_dd,
            'max_drawdown_pct': max_dd_pct,
            
            # Consecutivos
            'max_consecutive_wins': max_consecutive_wins,
            'max_consecutive_losses': max_consecutive_losses,
            
            # LONG vs SHORT
            'long_trades': len(long_trades),
            'long_win_rate': len(long_trades[long_trades['pnl'] > 0]) / len(long_trades) if len(long_trades) > 0 else 0,
            'long_pnl': long_trades['pnl'].sum() if len(long_trades) > 0 else 0,
            
            'short_trades': len(short_trades),
            'short_win_rate': len(short_trades[short_trades['pnl'] > 0]) / len(short_trades) if len(short_trades) > 0 else 0,
            'short_pnl': short_trades['pnl'].sum() if len(short_trades) > 0 else 0,
            
            # SL vs TP
            'sl_count': len(sl_trades),
            'tp_count': len(tp_trades),
            'sl_pct': len(sl_trades) / total_trades if total_trades > 0 else 0,
            'tp_pct': len(tp_trades) / total_trades if total_trades > 0 else 0,
        }
    
    def _calculate_sharpe(self, returns: np.ndarray) -> float:
        """Calcula Sharpe Ratio"""
        if len(returns) < 2:
            return 0.0
        
        mean_return = returns.mean()
        std_return = returns.std()
        
        if std_return == 0:
            return 0.0
        
        # Anualizar (assumindo trades diários, ajustar se necessário)
        sharpe = (mean_return - self.risk_free_rate) / std_return
        sharpe_annualized = sharpe * np.sqrt(252)  # 252 dias de trading
        
        return sharpe_annualized
    
    def _calculate_sortino(self, returns: np.ndarray) -> float:
        """Calcula Sortino Ratio (penaliza apenas downside)"""
        if len(returns) < 2:
            return 0.0
        
        mean_return = returns.mean()
        downside_returns = returns[returns < 0]
        
        if len(downside_returns) == 0:
            return 0.0
        
        downside_std = downside_returns.std()
        
        if downside_std == 0:
            return 0.0
        
        sortino = (mean_return - self.risk_free_rate) / downside_std
        sortino_annualized = sortino * np.sqrt(252)
        
        return sortino_annualized
    
    def _calculate_max_drawdown(self, equity_curve: np.ndarray) -> tuple:
        """Calcula Max Drawdown absoluto e percentual"""
        if len(equity_curve) == 0:
            return 0.0, 0.0
        
        # Adicionar capital inicial
        equity = 10000 + equity_curve
        
        # Running maximum
        running_max = np.maximum.accumulate(equity)
        
        # Drawdown
        drawdown = equity - running_max
        
        # Max drawdown absoluto
        max_dd = drawdown.min()
        
        # Max drawdown percentual
        max_dd_idx = drawdown.argmin()
        peak_value = running_max[max_dd_idx]
        max_dd_pct = (max_dd / peak_value) * 100 if peak_value != 0 else 0
        
        return max_dd, abs(max_dd_pct)
    
    def _max_consecutive(self, trades_df: pd.DataFrame, trade_type: str) -> int:
        """Calcula sequência máxima de wins ou losses"""
        if trade_type == 'win':
            mask = trades_df['pnl'] > 0
        else:
            mask = trades_df['pnl'] < 0
        
        max_streak = 0
        current_streak = 0
        
        for is_match in mask:
            if is_match:
                current_streak += 1
                max_streak = max(max_streak, current_streak)
            else:
                current_streak = 0
        
        return max_streak
    
    def _empty_metrics(self) -> Dict:
        """Retorna métricas vazias"""
        return {
            'total_return': 0,
            'total_return_pct': 0,
            'total_trades': 0,
            'winning_trades': 0,
            'losing_trades': 0,
            'win_rate': 0,
            'avg_win': 0,
            'avg_loss': 0,
            'avg_trade': 0,
            'sharpe_ratio': 0,
            'sortino_ratio': 0,
            'profit_factor': 0,
            'expectancy': 0,
            'max_drawdown': 0,
            'max_drawdown_pct': 0,
            'max_consecutive_wins': 0,
            'max_consecutive_losses': 0,
            'long_trades': 0,
            'long_win_rate': 0,
            'long_pnl': 0,
            'short_trades': 0,
            'short_win_rate': 0,
            'short_pnl': 0,
            'sl_count': 0,
            'tp_count': 0,
            'sl_pct': 0,
            'tp_pct': 0,
        }
    
    def calculate_composite_score(
        self, 
        metrics: Dict,
        weights: Optional[Dict] = None
    ) -> float:
        """
        Calcula score composto ponderado
        
        Args:
            metrics: Dict com métricas
            weights: Pesos personalizados (opcional)
        
        Returns:
            Score composto (0-100)
        """
        if weights is None:
            weights = {
                'sharpe': 0.30,
                'win_rate': 0.20,
                'profit_factor': 0.20,
                'total_return': 0.15,
                'max_drawdown': 0.15,
            }
        
        # Normalizar métricas (0-100)
        sharpe_norm = min(100, max(0, metrics['sharpe_ratio'] * 10))  # Sharpe 0-10 -> 0-100
        wr_norm = metrics['win_rate'] * 100  # 0-1 -> 0-100
        pf_norm = min(100, max(0, (metrics['profit_factor'] - 1) * 20))  # PF 1-6 -> 0-100
        ret_norm = min(100, max(0, metrics['total_return_pct']))  # Return direto
        dd_norm = max(0, 100 - metrics['max_drawdown_pct'])  # Inverter (menor DD = melhor)
        
        # Score ponderado
        score = (
            sharpe_norm * weights['sharpe'] +
            wr_norm * weights['win_rate'] +
            pf_norm * weights['profit_factor'] +
            ret_norm * weights['total_return'] +
            dd_norm * weights['max_drawdown']
        )
        
        return score

File: engines/rust/src/metrics.rs
use crate::types::{Metrics, Trade, TradeType, ExitReason};

pub fn calculate_metrics(trades: &[Trade]) -> Metrics {
    if trades.is_empty() {
        return Metrics::default();
    }

    let total_trades = trades.len();
    let returns: Vec<f32> = trades.iter().map(|t| t.pnl).collect();
    let total_return: f32 = returns.iter().sum();

    // Separar winners e losers
    let winning_trades: Vec<&Trade> = trades.iter().filter(|t| t.pnl > 0.0).collect();
    let losing_trades: Vec<&Trade> = trades.iter().filter(|t| t.pnl < 0.0).collect();

    let win_rate = winning_trades.len() as f32 / total_trades as f32;

    let avg_win = if !winning_trades.is_empty() {
        winning_trades.iter().map(|t| t.pnl).sum::<f32>() / winning_trades.len() as f32
    } else {
        0.0
    };

    let avg_loss = if !losing_trades.is_empty() {
        losing_trades.iter().map(|t| t.pnl).sum::<f32>() / losing_trades.len() as f32
    } else {
        0.0
    };

    let avg_trade = returns.iter().sum::<f32>() / total_trades as f32;

    // Profit Factor
    let gross_profit: f32 = winning_trades.iter().map(|t| t.pnl).sum();
    let gross_loss: f32 = losing_trades.iter().map(|t| t.pnl.abs()).sum();
    let profit_factor = if gross_loss > 0.0 {
        gross_profit / gross_loss
    } else {
        0.0
    };

    // Sharpe Ratio
    let sharpe_ratio = calculate_sharpe(&returns);

    // Sortino Ratio
    let sortino_ratio = calculate_sortino(&returns);

    // Max Drawdown
    let (max_dd, max_dd_pct) = calculate_max_drawdown(&returns);

    // Consecutive wins/losses
    let max_consecutive_wins = max_consecutive(trades, true);
    let max_consecutive_losses = max_consecutive(trades, false);

    // Expectancy
    let expectancy = (win_rate * avg_win) + ((1.0 - win_rate) * avg_loss);

    Metrics {
        total_return,
        total_return_pct: (total_return / 10000.0) * 100.0,
        total_trades,
        winning_trades: winning_trades.len(),
        losing_trades: losing_trades.len(),
        win_rate,
        avg_win,
        avg_loss,
        avg_trade,
        profit_factor,
        sharpe_ratio,
        sortino_ratio,
        max_drawdown: max_dd,
        max_drawdown_pct: max_dd_pct,
        max_consecutive_wins,
        max_consecutive_losses,
        expectancy,
    }
}

fn calculate_sharpe(returns: &[f32]) -> f32 {
    if returns.is_empty() {
        return 0.0;
    }

    let mean = returns.iter().sum::<f32>() / returns.len() as f32;
    let variance = returns.iter().map(|r| (r - mean).powi(2)).sum::<f32>() / returns.len() as f32;
    let std = variance.sqrt();

    if std > 0.0 {
        mean / std * (252.0_f32).sqrt()  // Anualized (252 trading days)
    } else {
        0.0
    }
}

fn calculate_sortino(returns: &[f32]) -> f32 {
    if returns.is_empty() {
        return 0.0;
    }

    let mean = returns.iter().sum::<f32>() / returns.len() as f32;
    
    // Downside deviation (apenas retornos negativos)
    let downside_returns: Vec<f32> = returns.iter().filter(|&&r| r < 0.0).copied().collect();
    
    if downside_returns.is_empty() {
        return 0.0;
    }

    let downside_variance = downside_returns.iter().map(|r| r.powi(2)).sum::<f32>() / downside_returns.len() as f32;
    let downside_std = downside_variance.sqrt();

    if downside_std > 0.0 {
        mean / downside_std * (252.0_f32).sqrt()
    } else {
        0.0
    }
}

fn calculate_max_drawdown(returns: &[f32]) -> (f32, f32) {
    if returns.is_empty() {
        return (0.0, 0.0);
    }

    // Calcular equity curve (cumsum)
    let mut equity_curve = Vec::with_capacity(returns.len());
    let mut cumsum = 0.0;
    for &ret in returns {
        cumsum += ret;
        equity_curve.push(cumsum);
    }

    let mut max_equity = f32::NEG_INFINITY;
    let mut max_dd = 0.0;

    for &equity in &equity_curve {
        if equity > max_equity {
            max_equity = equity;
        }
        let dd = max_equity - equity;
        if dd > max_dd {
            max_dd = dd;
        }
    }

    let max_dd_pct = if max_equity > 0.0 {
        (max_dd / max_equity) * 100.0
    } else {
        0.0
    };

    (max_dd, max_dd_pct)
}

fn max_consecutive(trades: &[Trade], wins: bool) -> usize {
    let mut max_count = 0;
    let mut current_count = 0;

    for trade in trades {
        let is_win = trade.pnl > 0.0;
        if is_win == wins {
            current_count += 1;
            if current_count > max_count {
                max_count = current_count;
            }
        } else {
            current_count = 0;
        }
    }

    max_count
}

